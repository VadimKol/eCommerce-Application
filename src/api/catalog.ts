import { toast } from 'react-toastify';

import { CurrencySymbols } from '@/common/enums';
import type { CategoriesData, Product, ProductDetails } from '@/common/types';
import { DEFAULT_LOCALE, getCheckboxFilter, getSort, QUERY_LIMIT } from '@/common/utils';

import { apiRoot } from './build-client';

export async function getCategories(): Promise<CategoriesData> {
  const categories: CategoriesData = [];
  try {
    const data = await apiRoot.categories().get().execute();

    data.body.results.forEach((category) => {
      if (!category.parent) {
        categories.push({
          id: category.id,
          name: category.name[DEFAULT_LOCALE] ?? '',
          slug: category.slug[DEFAULT_LOCALE] ?? '',
          key: category.key ?? '',
          subcategories: [],
        });
      }
    });

    data.body.results.forEach((subcategory) => {
      if (subcategory.parent) {
        const parentId = subcategory.parent.id;
        const category = categories.find((c) => c.id === parentId);
        if (category) {
          category.subcategories.push({
            id: subcategory.id,
            name: subcategory.name[DEFAULT_LOCALE] ?? '',
            slug: subcategory.slug[DEFAULT_LOCALE] ?? '',
            key: subcategory.key ?? '',
            products: [],
          });
        }
      }
    });
  } catch (error) {
    if (error instanceof Error) {
      toast(error.message, { type: 'error' });
    }
  }

  return [...categories];
}

export async function getProducts(
  offset: number,
  sortType: string,
  priceFilter: [number, number],
  franchises: boolean[],
  countriesF: boolean[],
  materialsF: boolean[],
  search: string,
  categoryID?: string,
  subcategoryID?: string,
): Promise<{ products: Product[]; total: number }> {
  const products: Product[] = [];
  let total = 0;
  const categoryRequest = subcategoryID ?? categoryID;
  const queryArgs = {
    limit: QUERY_LIMIT,
    offset: offset * 12,
    sort: getSort(sortType),
    filter: [`variants.price.centAmount:range (${priceFilter[0] * 100} to ${priceFilter[1] * 100})`],
  };
  if (categoryRequest) {
    queryArgs.filter.push(`categories.id:"${categoryRequest}"`);
  }
  const fandomsFilter = getCheckboxFilter(franchises, 'Fandom');
  if (fandomsFilter) {
    queryArgs.filter.push(fandomsFilter);
  }
  const countriesFilter = getCheckboxFilter(countriesF, 'Country');
  if (countriesFilter) {
    queryArgs.filter.push(countriesFilter);
  }
  const materialsFilter = getCheckboxFilter(materialsF, 'Material');
  if (materialsFilter) {
    queryArgs.filter.push(materialsFilter);
  }
  if (search) {
    Object.assign(queryArgs, { fuzzy: true, fuzzyLevel: 0, [`text.${DEFAULT_LOCALE}`]: search });
  }

  try {
    const data = await apiRoot.productProjections().search().get({ queryArgs }).execute();

    total = data.body.total ?? 0;

    data.body.results.forEach((product) => {
      const description = product.description ? product.description[DEFAULT_LOCALE] ?? '' : '';
      const { prices } = product.masterVariant;
      const price = prices && prices[0] ? (prices[0].value.centAmount / 100).toFixed(2) : '0.00';
      const discount =
        prices && prices[0] && prices[0].discounted ? (prices[0].discounted.value.centAmount / 100).toFixed(2) : '';
      const images = product.masterVariant.images ? product.masterVariant.images.map((image) => image.url) : [];
      const quantity = product.masterVariant.availability?.availableQuantity
        ? product.masterVariant.availability.availableQuantity
        : 0;
      products.push({
        id: product.id,
        name: product.name[DEFAULT_LOCALE] ?? '',
        description,
        slug: product.slug[DEFAULT_LOCALE] ?? '',
        key: product.key ?? '',
        sku: product.masterVariant.sku ?? '',
        quantity,
        price,
        discount,
        images,
        categories: product.categories.map(({ id }) => id),
        keyCategory: '',
        keySubCategory: '',
      });
    });
  } catch (error) {
    if (error instanceof Error) {
      toast(error.message, { type: 'error' });
    }
  }

  return { products: [...products], total };
}

export async function getProduct(productKey: string): Promise<ProductDetails> {
  const response = await apiRoot
    .productProjections()
    .get({
      queryArgs: {
        where: `key="${productKey}"`,
      },
    })
    .execute();

  const product = response.body.results[0];

  if (!product) {
    return Promise.reject(new Error('Product not found'));
  }

  const productTypeId = product.productType.id;

  const productTypeResponse = await apiRoot.productTypes().withId({ ID: productTypeId }).get().execute();

  const productType = productTypeResponse.body;

  const { id } = product;
  const name = product.name[DEFAULT_LOCALE];
  const description = product.description?.[DEFAULT_LOCALE];
  const { masterVariant } = product;
  const prices = masterVariant?.prices ?? [];
  const images = masterVariant?.images ?? [];
  const attributes = masterVariant?.attributes ?? [];
  const availability = masterVariant?.availability;

  const price = prices[0];
  const priceAmount = price?.value?.centAmount ? (price.value.centAmount / 100).toFixed(2) : null;
  const discountedAmount = price?.discounted ? (price.discounted.value.centAmount / 100).toFixed(2) : null;

  const currencySymbol = price?.value?.currencyCode
    ? CurrencySymbols[price.value.currencyCode as keyof typeof CurrencySymbols]
    : null;

  const attributeLabels = productType.attributes?.reduce(
    (acc, attr) => {
      acc[attr.name] = attr.label[DEFAULT_LOCALE]!;
      return acc;
    },
    {} as Record<string, string>,
  );

  return {
    id,
    name,
    description,
    price: priceAmount || null,
    discountedPrice: discountedAmount || null,
    currency: currencySymbol,
    images: images.map((image) => image.url),
    attributes: attributes.map((attr) => ({
      name: attributeLabels?.[attr.name] || attr.name,
      value: attr.value,
    })),
    availability: {
      isOnStock: availability?.isOnStock ?? null,
      availableQuantity: availability?.availableQuantity ?? null,
    },
  };
}
