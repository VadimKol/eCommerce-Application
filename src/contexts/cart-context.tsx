import type { Cart, LineItem } from '@commercetools/platform-sdk';
import { createContext, type ReactNode, useCallback, useEffect, useMemo, useState } from 'react';
import { toast } from 'react-toastify';

import { broadcastChannel } from '@/common/utils';

import {
  addPromocodeIntoCart,
  addToCart,
  changeQuantityFromCart,
  clearCart,
  createCart,
  getActiveCart,
  removeFromCart,
} from '../api/cart';

export type CartContextProps = {
  cart: Cart | null;
  isItemInCart: (id: string) => boolean;
  getItemCount: (id: string) => number;
  getLineItemId: (id: string) => string;
  addItemToCart: (id: string, quantity?: number) => Promise<void>;
  removeItemFromCart: (id: string, quantity?: number) => Promise<void>;
  getCartItemsCount: () => number | undefined;
  cartItems: LineItem[];
  updateCart: (updatedCard: Cart | null) => void;
  clearFromCart: (lineItems: string[]) => Promise<void>;
  changeItemQuantityFromCart: (id: string, quantity: number) => Promise<void>;
  addPromocodeToCart: (code: string) => Promise<void>;
};

export const CartContext = createContext({} as CartContextProps);

export function CartProvider({ children }: { children: ReactNode }): React.ReactElement {
  const [cart, setCart] = useState<Cart | null>(null);

  useEffect(() => {
    if (localStorage.getItem('geek-shop-refresh') === null) {
      createCart()
        .then((response) => setCart(response.body))
        .catch(() => toast(`Failed to create cart`, { type: 'error' }));
    } else {
      getActiveCart()
        .then((response) => setCart(response.body))
        .catch(() => toast(`Failed to get cart`, { type: 'error' }));
    }
    const handleCart = (
      e: MessageEvent<{
        type: string;
        payload: { cart: Cart | null };
      }>,
    ): void => {
      if (e.data.type === 'Password' || e.data.type === 'Auth' || e.data.type === 'Cart') {
        setCart(e.data.payload.cart);
      }
    };
    broadcastChannel.addEventListener('message', handleCart);
    return (): void => {
      broadcastChannel.removeEventListener('message', handleCart);
    };
  }, []);

  const updateCart = useCallback((updatedCard: Cart | null): void => {
    setCart(updatedCard);
    broadcastChannel.postMessage({ type: 'Cart', payload: { cart: updatedCard } });
  }, []);

  const cartItems = useMemo(() => (cart ? cart.lineItems : []), [cart]);

  const isItemInCart = useCallback(
    (id: string): boolean => cartItems.some((item) => item.productId === id),
    [cartItems],
  );

  const getLineItemId = useCallback(
    (id: string): string => {
      const item = cartItems.find((cartItem) => cartItem.productId === id);
      return item ? item.id : '';
    },
    [cartItems],
  );

  const getItemCount = useCallback(
    (id: string): number => {
      const item = cartItems.find((cartItem) => cartItem.productId === id);
      return item ? item.quantity : 0;
    },
    [cartItems],
  );

  const addItemToCart = useCallback(
    async (id: string, quantity?: number): Promise<void> => {
      const response = await addToCart(cart!, id, quantity);

      setCart(response.body);
      broadcastChannel.postMessage({ type: 'Cart', payload: { cart: response.body } });
    },
    [cart],
  );

  const removeItemFromCart = useCallback(
    async (id: string, quantity?: number): Promise<void> => {
      const lineItemId = getLineItemId(id);

      if (lineItemId) {
        const response = await removeFromCart(cart!, lineItemId, quantity);

        setCart(response.body);
        broadcastChannel.postMessage({ type: 'Cart', payload: { cart: response.body } });
      }
    },
    [cart, getLineItemId],
  );

  const getCartItemsCount = useCallback(
    (): number => cartItems.reduce((count, item) => count + item.quantity, 0),
    [cartItems],
  );

  const clearFromCart = useCallback(
    async (lineItems: string[]): Promise<void> => {
      const response = await clearCart(cart!, lineItems);

      setCart(response.body);
      broadcastChannel.postMessage({ type: 'Cart', payload: { cart: response.body } });
    },
    [cart],
  );

  const changeItemQuantityFromCart = useCallback(
    async (id: string, quantity: number): Promise<void> => {
      const lineItemId = getLineItemId(id);

      if (lineItemId) {
        const response = await changeQuantityFromCart(cart!, lineItemId, quantity);

        setCart(response.body);
        broadcastChannel.postMessage({ type: 'Cart', payload: { cart: response.body } });
      }
    },
    [cart, getLineItemId],
  );

  const addPromocodeToCart = useCallback(
    async (code: string): Promise<void> => {
      const response = await addPromocodeIntoCart(cart!, code);

      setCart(response.body);
      broadcastChannel.postMessage({ type: 'Cart', payload: { cart: response.body } });
    },
    [cart],
  );

  const CartContextValue: CartContextProps = useMemo(
    () => ({
      cart,
      addItemToCart,
      removeItemFromCart,
      isItemInCart,
      getItemCount,
      getLineItemId,
      cartItems,
      getCartItemsCount,
      updateCart,
      clearFromCart,
      changeItemQuantityFromCart,
      addPromocodeToCart,
    }),
    [
      addItemToCart,
      removeItemFromCart,
      isItemInCart,
      getItemCount,
      getLineItemId,
      cartItems,
      getCartItemsCount,
      cart,
      updateCart,
      clearFromCart,
      changeItemQuantityFromCart,
      addPromocodeToCart,
    ],
  );

  return <CartContext.Provider value={CartContextValue}>{children}</CartContext.Provider>;
}
