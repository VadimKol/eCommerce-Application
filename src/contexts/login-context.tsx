import { createContext, useCallback, useMemo, useState } from 'react';

export interface LoginContextInterface {
  isAuthenticated: boolean;
  setIsAuthenticated: React.Dispatch<React.SetStateAction<boolean>>;
  handleLogout: () => void;
}

export const LoginContext = createContext({
  isAuthenticated: false,
  setIsAuthenticated: () => {},
  handleLogout: () => {},
} as LoginContextInterface);

export function LoginProvider({ children }: { children: React.ReactNode }): JSX.Element {
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  const handleLogout = useCallback(() => {
    setIsAuthenticated(false);
  }, []);

  const contextValue = useMemo(
    () => ({ isAuthenticated, setIsAuthenticated, handleLogout }),
    [isAuthenticated, handleLogout],
  );

  return <LoginContext.Provider value={contextValue}>{children}</LoginContext.Provider>;
}
