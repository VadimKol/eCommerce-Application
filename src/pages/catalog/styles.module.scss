.container {
  display: flex;
  flex-direction: column;
  gap: 10px;
  box-sizing: border-box;
  width: 100%;
  padding: 0 20px;
  text-align: center;
  color: var(--primary-color);
}

.title {
  margin-block: 0;
}

.text {
  margin-block: 0;
}

.product_box {
  display: flex;
  gap: 15px;
  justify-content: center;
  margin-bottom: 30px;
}

.products {
  display: flex;
  flex-wrap: wrap;
  gap: 15px;
  width: 75%;
  margin: 0;
  padding: 0;
  list-style-type: none;
}

.search_form {
  display: flex;
  justify-content: flex-end;

  & .search_box {
    position: relative;
  }

  & .search {
    box-sizing: border-box;
    width: 250px;
    height: 30px;
    padding: 0 30px 0 5px;
    font-family: var(--primary-font);
    border: 1px solid var(--primary-color);
    border-radius: 6px;

    &:focus {
      border: 1px solid #000000;
    }

    &:focus-visible {
      outline: none;
    }
  }

  & .search_button {
    position: absolute;
    top: 6px;
    right: 6px;
    width: 20px;
    height: 20px;
    padding: 0;
    background-color: transparent;
    background-image: url("/src/assets/images/search.svg");
    background-size: cover;
    border: none;


    @media (hover: hover) {
      &:hover {
        cursor: pointer;
      }
    }
  }
}

.sort_form {
  display: flex;
  justify-content: flex-end;
}

.product_img {
  transition: transform 0.5s;
}

.product_link {
  display: inline-block;
  max-width: 375px;
  padding: 30px;
  text-decoration: none;
  color: var(--primary-color);
  background:
    linear-gradient(var(--background-color) 0 0) padding-box,
    linear-gradient(45deg, var(--primary-color), var(--secondary-color)) border-box;
  border: 5px solid transparent;
  border-radius: 10px;

  @media (hover: hover) {
    &:hover {
      .product_img {
        transform: scale(1.1);
      }
    }
  }
}

.product_description {
  max-width: 300px;
  text-align: left;
}

.product_price {
  margin-bottom: 0;
  font-weight: 600;
  text-align: left;
}

.filters {
  width: 20%;
  height: fit-content;
  font-size: 1.5em;
  font-weight: 600;
  background-color: var(--background-color);
}
