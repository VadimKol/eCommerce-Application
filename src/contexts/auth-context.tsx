import { createContext, useCallback, useEffect, useMemo, useState } from 'react';

import { apiRoot, RefreshTokenFlow } from '@/api/build-client';
import type { AuthContextInterface } from '@/common/types';
import { broadcastChannel } from '@/common/utils';

export const AuthContext = createContext({
  isAuthenticated: false,
  handleLogin: () => {},
  handleLogout: () => {},
} as AuthContextInterface);

export function AuthProvider({ children }: { children: React.ReactNode }): JSX.Element {
  const [isAuthenticated, setIsAuthenticated] = useState(localStorage.getItem('geek-shop-auth') !== null);

  const handleLogin = useCallback(() => {
    setIsAuthenticated(true);
  }, []);

  const handleLogout = useCallback(() => {
    setIsAuthenticated(false);
  }, []);

  const contextValue = useMemo(
    () => ({ isAuthenticated, handleLogin, handleLogout }),
    [isAuthenticated, handleLogout, handleLogin],
  );

  useEffect(() => {
    const handleAuth = (e: MessageEvent<{ type: string }>): void => {
      if (e.data.type === 'Auth') {
        if (!isAuthenticated) {
          handleLogin();
        } else {
          handleLogout();
        }
      }
      if (e.data.type === 'Password' || e.data.type === 'Auth') {
        Object.assign(apiRoot, RefreshTokenFlow(localStorage.getItem('geek-shop-refresh') || ''));
      }
    };
    broadcastChannel.addEventListener('message', handleAuth);
    return (): void => {
      broadcastChannel.removeEventListener('message', handleAuth);
    };
  }, [isAuthenticated, handleLogin, handleLogout]);

  return <AuthContext.Provider value={contextValue}>{children}</AuthContext.Provider>;
}
