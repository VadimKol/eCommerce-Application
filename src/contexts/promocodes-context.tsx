import { createContext, type ReactNode, useEffect, useMemo, useState } from 'react';
import { toast } from 'react-toastify';

import { getPromocodes } from '@/api/cart';
import { DEFAULT_LOCALE } from '@/common/utils';

export interface Promocode {
  id: string;
  name: string;
}

export type PromocodesProps = {
  promocodes: Promocode[] | null;
};

export const PromocodesContext = createContext({} as PromocodesProps);

export function PromocodesProvider({ children }: { children: ReactNode }): React.ReactElement {
  const [promocodes, setPromocodes] = useState<Promocode[] | null>(null);

  useEffect(() => {
    getPromocodes()
      .then((response) =>
        setPromocodes(
          response.body.results.map((promocode) => ({
            id: promocode.id,
            name: promocode.name ? promocode.name[DEFAULT_LOCALE] || '' : '',
          })),
        ),
      )
      .catch(() => toast(`Failed to get promocodes`, { type: 'error' }));
  }, []);

  const CartContextValue: PromocodesProps = useMemo(
    () => ({
      promocodes,
    }),
    [promocodes],
  );

  return <PromocodesContext.Provider value={CartContextValue}>{children}</PromocodesContext.Provider>;
}
