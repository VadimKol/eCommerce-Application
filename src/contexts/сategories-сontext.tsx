import type { ReactNode } from 'react';
import { createContext, useEffect, useMemo, useState } from 'react';
import { useLocation } from 'react-router-dom';

import { getCategories } from '@/api/catalog';
import { NavigationPaths } from '@/common/enums';
import type { CategoriesContextType, CategoriesData } from '@/common/types';
import { StatusError } from '@/common/utils';

export const CategoriesContext = createContext<CategoriesContextType | undefined>(undefined);

type CategoriesProviderProps = {
  children: ReactNode;
};

let categoriesCache: CategoriesData | null = null;

export function CategoriesProvider({ children }: CategoriesProviderProps): JSX.Element {
  const [categories, setCategories] = useState<CategoriesData | null>(categoriesCache);
  const [loading, setLoading] = useState(!categoriesCache);
  const [error, setError] = useState<Error | null>(null);
  const location = useLocation();

  const validatePath = (categoriesData: CategoriesData, pathname: string): void => {
    const pathParts = pathname.split('/').filter(Boolean);

    if (pathParts.length < 2 || `/${pathParts[0]}` !== NavigationPaths.CATALOG.toString()) {
      return;
    }

    const categoryKey = pathParts[1];
    const category = categoriesData.find((catData) => catData.key === categoryKey);

    if (!category) {
      setError(new StatusError('Category not found', 404));
      return;
    }

    if (pathParts.length > 2) {
      const subcategoryKey = pathParts[2];
      const subcategory = category.subcategories.find((subData) => subData.key === subcategoryKey);

      if (!subcategory) {
        setError(new StatusError('Subcategory not found', 404));
      }
    }
  };

  useEffect(() => {
    setError(null);

    if (categoriesCache) {
      setLoading(false);
      setCategories(categoriesCache);
      validatePath(categoriesCache, location.pathname);
      return;
    }

    if (!location.pathname.startsWith(NavigationPaths.CATALOG)) {
      return;
    }

    getCategories()
      .then((data) => {
        categoriesCache = data;
        setCategories(data);
        validatePath(data, location.pathname);
      })
      .catch((err) => {
        if (err instanceof Error) {
          setError(err);
        } else {
          setError(new Error('An unknown error occurred'));
        }
      })
      .finally(() => {
        setLoading(false);
      });
  }, [location.pathname]);

  const value = useMemo(() => ({ categories, loading, error }), [categories, loading, error]);

  return <CategoriesContext.Provider value={value}>{children}</CategoriesContext.Provider>;
}
