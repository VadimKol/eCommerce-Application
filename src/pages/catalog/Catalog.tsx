import classNames from 'classnames';
import { useEffect, useRef, useState } from 'react';
import { useParams } from 'react-router-dom';
import { toast } from 'react-toastify';

import { getProducts } from '@/api/client-actions';
import type { Product } from '@/common/types';
import { fandoms, PRICE_FILTER_MAX, PRICE_FILTER_MIN, QUERY_LIMIT, sortingTypes, StatusError } from '@/common/utils';
import { Breadcrumbs } from '@/components/breadcrumbs/Breadcrumbs';
import { CategoriesList } from '@/components/categories-list/CategoriesList';
import { CustomSelect } from '@/components/custom-select/CustomSelect';
import { Filters } from '@/components/filters/Filters';
import { ProductCard } from '@/components/product-card/ProductCard';
import { useCategories } from '@/hooks/useCategories';
import { NoMatch } from '@/pages/no-match/NoMatch';

import styles from './styles.module.scss';

export function Catalog(): JSX.Element {
  const { error, categories: categoriesData, loading } = useCategories();
  const [products, setProducts] = useState<Product[]>([]);
  const { categoryName, subcategoryName } = useParams<{ categoryName?: string; subcategoryName?: string }>();
  const [categories, setCategories] = useState({ categoryName, subcategoryName });
  const [page, setPage] = useState(0);
  const [sortType, setSortType] = useState('Sort');
  const [search, setSearch] = useState('');
  const searchField = useRef<HTMLInputElement>(null);

  const total = useRef(0);

  const [priceFilter, setPriceFilter] = useState<[number, number]>([PRICE_FILTER_MIN, PRICE_FILTER_MAX]);
  const [franchises, setFranchises] = useState(Array<boolean>(fandoms.length).fill(false));

  const [loadingProducts, setLoadingProducts] = useState(false);

  if (categories.categoryName !== categoryName || categories.subcategoryName !== subcategoryName) {
    setCategories({ categoryName, subcategoryName });
    setPage(0);
    setSortType('Sort');
    setPriceFilter([PRICE_FILTER_MIN, PRICE_FILTER_MAX]);
    setFranchises(Array(fandoms.length).fill(false));
    setSearch('');
    if (searchField.current?.value) {
      searchField.current.value = '';
    }
  }

  useEffect(() => {
    if (categoriesData) {
      setLoadingProducts(true);
      const categoryID = categoriesData?.find((category) => category.key === categories.categoryName);
      const subcategoryID = categoryID?.subcategories.find(
        (subcategory) => subcategory.key === categories.subcategoryName,
      );
      getProducts(page, sortType, priceFilter, franchises, search, categoryID?.id, subcategoryID?.id)
        .then((data) => {
          total.current = data.total;
          data.products.forEach((product) => {
            const category = categoriesData?.find((c) => c.id === product.categoryId);
            if (category) {
              product.slugCategory = category.slug;
              product.keyCategory = category.key;
            }
            const subcategory = category?.subcategories.find((sc) => sc.id === product.subcategoryId);
            if (subcategory) {
              product.slugSubCategory = subcategory.slug;
              product.keySubCategory = subcategory.key;
            }
          });
          setProducts(data.products);
        })
        .catch((e: Error) => toast(e.message, { type: 'error' }))
        .finally(() => setLoadingProducts(false));
    }
  }, [categories, categoriesData, page, sortType, priceFilter, franchises, search]);

  if (error instanceof StatusError && error.statusCode === 404) {
    return <NoMatch />;
  }

  if (loading) {
    return (
      <main className={classNames('main', styles.main)}>
        <div className={styles.infoContainer}>Loading data...</div>
      </main>
    );
  }

  if (!categoriesData) {
    return (
      <main className={classNames('main', styles.main)}>
        <div className={styles.infoContainer}>No categories data</div>
      </main>
    );
  }

  return (
    <main className="main">
      <div className={styles.topContainer}>
        <CategoriesList parentClass={styles.categoriesList} />
      </div>
      <Breadcrumbs parentClass={styles.breadcrumbs} />
      <div className={styles.container}>
        <section className={styles.product_box}>
          <aside className={styles.filters}>
            <Filters
              priceFilter={priceFilter}
              setPriceFilter={setPriceFilter}
              franchises={franchises}
              setFranchises={setFranchises}
              setPage={setPage}
            />
          </aside>
          {loadingProducts ? (
            <div className={styles.infoContainer}>Loading products...</div>
          ) : (
            <section className={styles.products_section}>
              <div className={styles.sort_n_search_form}>
                <form
                  className={styles.search_form}
                  onSubmit={(e) => {
                    e.preventDefault();
                    if (typeof searchField.current?.value === 'string') {
                      setSearch(searchField.current?.value.trim());
                      setPage(0);
                    }
                  }}
                >
                  <div className={styles.search_box}>
                    <input
                      id="search"
                      className={styles.search}
                      type="text"
                      placeholder="Search..."
                      ref={searchField}
                    />
                    <button type="submit" className={styles.search_button} aria-label="Search-button" />
                  </div>
                </form>
                <CustomSelect selectItems={sortingTypes} selectState={sortType} setSelectState={setSortType} />
              </div>
              <ul className={products.length ? styles.products : `${styles.products} ${styles.products_not_found}`}>
                {products.map((product) => (
                  <li key={product.id} className={styles.products_item}>
                    <ProductCard product={product} categoryName={categoryName} subcategoryName={subcategoryName} />
                  </li>
                ))}
                {!products.length && <li className={styles.not_found}>Products not found</li>}
              </ul>
              {total.current > QUERY_LIMIT && (
                <div className={styles.pagination}>
                  {Boolean(page) && (
                    <button
                      type="button"
                      aria-label="Left"
                      className={styles.pag_left}
                      onClick={() => {
                        setPage(page - 1);
                      }}
                    />
                  )}
                  <div className={styles.pages}>
                    <span>{page + 1}</span>-
                    <span>{Math.floor(total.current / QUERY_LIMIT) + (total.current % QUERY_LIMIT > 0 ? 1 : 0)}</span>
                  </div>
                  {page + 1 < Math.floor(total.current / QUERY_LIMIT) + (total.current % QUERY_LIMIT > 0 ? 1 : 0) && (
                    <button
                      type="button"
                      aria-label="Right"
                      className={styles.pag_right}
                      onClick={() => {
                        setPage(page + 1);
                      }}
                    />
                  )}
                </div>
              )}
            </section>
          )}
        </section>
      </div>
    </main>
  );
}
