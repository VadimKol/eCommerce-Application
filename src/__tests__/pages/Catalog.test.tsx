import { render, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { CategoriesProvider } from '@/contexts/сategories-сontext';
import { Catalog } from '@/pages/catalog/Catalog';

it('renders correctly', async () => {
  const { container } = render(
    <BrowserRouter>
      <CategoriesProvider>
        <Catalog />
      </CategoriesProvider>
    </BrowserRouter>,
  );
  await waitFor(() => {
    expect(container).toMatchSnapshot();
  });
});
