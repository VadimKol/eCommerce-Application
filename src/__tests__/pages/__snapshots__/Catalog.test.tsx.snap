// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <main
    class="main main"
  >
    <div
      class="infoContainer"
    >
      Loading products...
    </div>
  </main>
</div>
`;
